File: backend/app/modules/recruitment/models/departure.py
"""Database models for volunteer departure interviews."""

from datetime import date, datetime
from typing import TYPE_CHECKING

from sqlalchemy import JSON, Boolean, Date, DateTime, ForeignKey, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.infrastructure.sql.base import Base

if TYPE_CHECKING:
    from app.modules.pi.models.volunteer import Volunteer


class DepartureField(Base):
    __tablename__ = "departure_fields"

    id: Mapped[int] = mapped_column(primary_key=True)
    key: Mapped[str] = mapped_column(String(100), unique=True, index=True)
    label: Mapped[str] = mapped_column(String(250))
    field_type: Mapped[str] = mapped_column(String(30), default="text")
    required: Mapped[bool] = mapped_column(Boolean, default=False)
    placeholder: Mapped[str] = mapped_column(String(250), default="")
    options: Mapped[list] = mapped_column(JSON, default=list)
    position: Mapped[int] = mapped_column(default=0, index=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    is_system: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )


class DepartureInterview(Base):
    __tablename__ = "departure_interviews"

    id: Mapped[int] = mapped_column(primary_key=True)
    volunteer_id: Mapped[int] = mapped_column(
        ForeignKey("volunteers.id", ondelete="RESTRICT"),
        nullable=False,
        unique=True,
        index=True,
    )
    departure_date: Mapped[date] = mapped_column(Date, nullable=False, index=True)
    departure_reason: Mapped[str] = mapped_column(Text)
    stay_in_contact: Mapped[bool] = mapped_column(Boolean, default=False)
    answers: Mapped[list] = mapped_column(JSON, default=list)
    completed_by_id: Mapped[int | None] = mapped_column(nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )
    volunteer: Mapped["Volunteer"] = relationship()

File: frontend/src/types/departure.ts
import type { RecruitmentFieldDraft, RecruitmentFieldType } from './recruitment';

export interface DepartureFieldDraft extends RecruitmentFieldDraft {
  field_type: Exclude<RecruitmentFieldType, 'email' | 'tel'>;
}

export interface DepartureField extends DepartureFieldDraft {
  id: number;
  key: string;
  position: number;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface DepartureAnswer {
  key: string;
  label: string;
  field_type: string;
  value: unknown;
}

export interface DepartureInterview {
  id: number;
  volunteer_id: number;
  departure_date: string;
  departure_reason: string;
  stay_in_contact: boolean;
  answers: DepartureAnswer[];
  completed_by_id: number | null;
  created_at: string;
  updated_at: string;
  volunteer: {
    id: number;
    full_name: string;
    email: string;
  };
}

export interface DepartureSelfService {
  volunteer: DepartureInterview['volunteer'];
  fields: DepartureField[];
  interview: DepartureInterview | null;
}
